extern/cfe/src/bmi_cfe.c

```diff
@@ -778,123 +804,67 @@ static int Get_adjusted_index_for_variable(const char *name)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-/* Uniform rectilinear */
-static int Get_grid_shape(Bmi *self, int grid, int *shape)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_spacing(Bmi *self, int grid, double *spacing)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_origin(Bmi *self, int grid, double *origin)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-/* Non-uniform rectilinear, curvilinear */
-static int Get_grid_x(Bmi *self, int grid, double *x)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_y(Bmi *self, int grid, double *y)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_z(Bmi *self, int grid, double *z)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_node_count(Bmi *self, int grid, int *count)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_edge_count(Bmi *self, int grid, int *count)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_face_count(Bmi *self, int grid, int *count)
-{
-    return BMI_FAILURE;
-}
-
-
-// TODO: complete implementation
-static int Get_grid_edge_nodes(Bmi *self, int grid, int *edge_nodes)
+static int Get_grid_rank (Bmi *self, int grid, int * rank)
 {
-    return BMI_FAILURE;
+    if (grid == 0) {
+        *rank = 1;
+        return BMI_SUCCESS;
+    }
+    else {
+        *rank = -1;
+        return BMI_FAILURE;
+    }
 }
 
 
-// TODO: complete implementation
-static int Get_grid_face_edges(Bmi *self, int grid, int *face_edges)
+static int Get_grid_size(Bmi *self, int grid, int * size)
 {
-    return BMI_FAILURE;
+    if (grid == 0) {
+        *size = 1;
+        return BMI_SUCCESS;
+    }
+    else {
+        *size = -1;
+        return BMI_FAILURE;
+    }
 }
 
 
-// TODO: complete implementation
-static int Get_grid_face_nodes(Bmi *self, int grid, int *face_nodes)
+static int Get_grid_type (Bmi *self, int grid, char * type)
 {
-    return BMI_FAILURE;
-}
+    int status = BMI_FAILURE;
 
-
-// TODO: complete implementation
-static int Get_grid_nodes_per_face(Bmi *self, int grid, int *nodes_per_face)
-{
-    return BMI_FAILURE;
+    if (grid == 0) {
+        strncpy(type, "scalar", BMI_MAX_TYPE_NAME);
+        status = BMI_SUCCESS;
+    }
+    else {
+        type[0] = '\0';
+        status = BMI_FAILURE;
+    }
+    return status;
 }
 
 
-// TODO: complete implementation
 static int Get_var_grid(Bmi *self, const char *name, int *grid)
 {
+
+    // Check to see if in output array first
+    for (int i = 0; i < OUTPUT_VAR_NAME_COUNT; i++) {
+        if (strcmp(name, output_var_names[i]) == 0) {
+            *grid = output_var_grids[i];
+            return BMI_SUCCESS;
+        }
+    }
+    // Then check to see if in input array
```

Comment:
Is a variable ever possible an output AND and input variable?

Reply:
Yes, it is possible.  In fact many of the [BMI examples](https://github.com/csdms/bmi-example-c/blob/master/heat/bmi_heat.c) have a single variable defined as both input and output.  Don't see this separation being problematic for CFE as, 

- There are no true bmi input variables
- & even if there was an identical i/o var, they would share the same "information characteristics" (type, units, grid (id), grid_type, location, etc.)